service/storage_proxy.cc

```diff
@@ -4294,7 +4296,7 @@ class digest_read_resolver : public abstract_read_resolver {
         if (waiting_for(ep)) {
             _failed++;
         }
-        if (kind == error_kind::DISCONNECT && _block_for == _target_count_for_cl) {
+        if (_speculating && kind == error_kind::DISCONNECT && _block_for == _target_count_for_cl) {
```

Comment:
Shouldn't it be a failure allowance? If we're speculating, we can allow one failure. If we're not speculating, we allow zero failures.

So it's more accurate to check for _failed < _allowed_failures, and select _allowed_failures according to speculation.

Reply:
With the adjustment, two failures in a speculating read will fall through and fail the request.

Reply:
Failure allowance seems to be there already -- below this "if":

```
        if (_block_for + _failed > _target_count_for_cl) {
            // fail_request() depending on the fail kind
        }
```

Reply:
Then maybe we should use it here instead of adding _speculating?